src/linear_solvers/custom_matrix.f90

```diff
@@ -10,39 +10,95 @@ module custom_matrix
     real(ccs_real), dimension(:), allocatable :: values
     integer(ccs_int), dimension(:), allocatable :: columns
     integer(ccs_int) :: values_per_row
+    integer(ccs_int) :: num_rows
   end type csr_matrix
 
 contains
 
-  subroutine create_new_matrix(rows, values_per_row, matrix)
+  subroutine create_new_matrix(num_rows, values_per_row, matrix)
 
-    integer(ccs_int), intent(in) :: rows
+    integer(ccs_int), intent(in) :: num_rows
     integer(ccs_int), intent(in) :: values_per_row
     type(csr_matrix), intent(inout) :: matrix
 
     matrix%values_per_row = values_per_row
+    matrix%num_rows = num_rows
 
-    allocate(matrix%values(rows * values_per_row))
+    allocate(matrix%values(num_rows * values_per_row))
     matrix%values(:) = 0.0
 
-    allocate(matrix%columns(rows * values_per_row))
+    allocate(matrix%columns(num_rows * values_per_row))
     matrix%columns(:) = 0
 
   end subroutine
 
+  subroutine get_offsets(row, start, end, matrix)
+    integer(ccs_int), intent(in) :: row
+    integer(ccs_int), intent(out) :: start
+    integer(ccs_int), intent(out) :: end
+    type(csr_matrix), intent(in) :: matrix
+
+    start = (row - 1) * matrix%values_per_row + 1
+    end = start + matrix%values_per_row
```

Comment:
Is `end` is off by 1?
e.g. for a matrix with 4 columns and 3 rows you'd get:
1:5
5:9
9:13

Reply:
yes, you seem to be right. Though it is consistent with the previous code:
```
start = offset
end = offset + matrix%values_per_row
```
So something probably worth checking with the existing use of insert_values

Reply:
Ah yes, sorry. I forgot I changed it in the `use_csr_omp` branch.